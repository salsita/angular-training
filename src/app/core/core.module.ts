import { HttpClientModule } from '@angular/common/http';
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { ApiModule } from '@salsita/ng-api';
import { CrudModule } from '@salsita/ng-crud';
import { EntityRepositoryModule } from '@salsita/ng-entities';
import { RouterModule as RouterStoreModule } from '@salsita/ng-router';

import { environment } from '../../environments/environment';
import { SharedModule } from '../shared/shared.module';
import { ApiErrorToastComponent } from './api/api-error-toast/api-error-toast.component';
import { ApiLoaderComponent } from './api/api-loader/api-loader.component';

@NgModule({
  imports: [
    HttpClientModule,

    ApiModule.forRoot({ storeNamespace: 'api', baseUrl: environment.apiBaseUrl }),
    CrudModule.forRoot({ storeNamespace: 'crud' }),
    EntityRepositoryModule.forRoot({ storeNamespace: 'entityRepository' }),
    RouterStoreModule.forRoot({ storeNamespace: 'router' }),
    SharedModule
  ],
  declarations: [ApiErrorToastComponent, ApiLoaderComponent],
  exports: [ApiErrorToastComponent, ApiLoaderComponent, ApiModule, HttpClientModule]
})
export class CoreModule {
  constructor(
    @Optional()
    @SkipSelf()
    parentModule: CoreModule
  ) {
    if (parentModule) {
      throw new Error(
        `CoreModule has already been loaded. Import Core modules in the AppModule only.`
      );
    }
  }
}
