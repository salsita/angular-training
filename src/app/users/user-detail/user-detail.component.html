<app-user-form
  *ngIf="user$ | async as user; else loading"
  [user]="user"
  [skills]="skills$ | async"
  (submit)="onSave($event)"
></app-user-form>
<ng-template #loading>
  Loading user...
</ng-template>
