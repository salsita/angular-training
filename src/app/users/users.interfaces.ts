export interface Skill {
  id: string;
  skill: string;
}

export interface UserSkill {
  skill: Skill;
}

export interface User {
  id: string;
  firstName: string;
  lastName: string;
  email: string;
  skills: UserSkill[];
}
