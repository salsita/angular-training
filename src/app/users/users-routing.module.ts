import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';

import { SkillsResolver } from './skills/skills.resolver';
import { UserCreateComponent } from './user-create/user-create.component';
import { UserDetailComponent } from './user-detail/user-detail.component';
import { UserDetailResolver } from './user-detail/user-detail.resolver';
import { UsersListComponent } from './users-list/users-list.component';
import { UsersListResolver } from './users-list/users-list.resolver';


const routes: Routes = [
  {
    path: '',
    resolve: {
      skills: SkillsResolver
    },
    children: [
      {
        path: '',
        component: UsersListComponent,
        resolve: {
          users: UsersListResolver
        },
        children: [
          {
            path: 'create',
            component: UserCreateComponent
          }
        ]
      },
      {
        path: ':userId',
        component: UserDetailComponent,
        resolve: {
          user: UserDetailResolver
        }
      }
    ]
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule],
  providers: []
})
export class UsersRoutingModule { }
