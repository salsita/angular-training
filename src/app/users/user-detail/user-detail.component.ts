import { ChangeDetectionStrategy, Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs/Observable';
import { usersActionCreators } from '../+state/users.actions';
import { getSkills, getUserDetail } from '../+state/users.selectors';
import { Skill, User } from '../users.interfaces';

@Component({
  selector: 'app-user-detail',
  templateUrl: './user-detail.component.html',
  styleUrls: ['./user-detail.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class UserDetailComponent {
  skills$: Observable<Skill[]>;
  user$: Observable<User | null>;

  constructor(private store: Store<any>) {
    this.user$ = getUserDetail(store);
    this.skills$ = getSkills(store);
  }

  onSave(user: User) {
    this.store.dispatch(usersActionCreators.saveUser(user));
  }
}
