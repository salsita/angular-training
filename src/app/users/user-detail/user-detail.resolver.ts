import { Injectable } from '@angular/core';
import { CrudResolver, CrudService } from '@salsita/ng-crud';
import { FlatRouterStateSnapshot } from '@salsita/ng-router';

import { userSchema } from '../+state/users.schema';
import { UsersApi } from '../users.api';

@Injectable()
export class UserDetailResolver extends CrudResolver {
  blocking = false;
  schema = userSchema;
  route = 'users/detail';
  key = 'user';

  constructor(private api: UsersApi, crud: CrudService) {
    super(crud);
  }

  params(route: FlatRouterStateSnapshot) {
    return [route.params.userId];
  }

  data(userId: string) {
    return this.api.getUser(userId);
  }
}
