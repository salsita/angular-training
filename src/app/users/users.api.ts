import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';

import { apiCall } from '../core/api/api.decorators';
import { CrudId } from '../core/crud/crud.interfaces';
import { User } from './users.interfaces';

@Injectable()
export class UsersApi {
  constructor(private http: HttpClient) {}

  @apiCall()
  getUsers() {
    return this.http.get(`/users`);
  }

  @apiCall()
  getUser(id: CrudId) {
    return this.http.get(`/users/${id}`);
  }

  @apiCall()
  deleteUser(id: CrudId) {
    return this.http.delete(`/users/${id}`);
  }

  @apiCall()
  saveUser({ id, ...data }: User) {
    return id ? this.http.patch(`/users/${id}`, data) : this.http.post('/users', data);
  }

  @apiCall()
  getSkills() {
    return this.http.get(`/skills`);
  }
}
