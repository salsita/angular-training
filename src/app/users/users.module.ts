import { NgModule } from '@angular/core';
import { EffectsModule } from '@ngrx/effects';

import { SharedModule } from '../shared/shared.module';
import { UsersEffects } from './+state/users.effects';
import { UsersSelectors } from './+state/users.selectors';
import { UserCreateComponent } from './user-create/user-create.component';
import { UserDetailComponent } from './user-detail/user-detail.component';
import { UserFormComponent } from './user-form/user-form.component';
import { UsersListComponent } from './users-list/users-list.component';
import { UsersRoutingModule } from './users-routing.module';
import { UsersApi } from './users.api';

@NgModule({
  imports: [UsersRoutingModule, EffectsModule.forFeature([UsersEffects]), SharedModule],
  declarations: [UsersListComponent, UserDetailComponent, UserFormComponent, UserCreateComponent],
  providers: [UsersApi, UsersEffects, UsersSelectors]
})
export class UsersModule {}
