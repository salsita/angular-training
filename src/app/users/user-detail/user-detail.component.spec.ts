import { NO_ERRORS_SCHEMA } from '@angular/core';
import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { Store } from '@ngrx/store';
import { MockStore } from '@salsita/ngrx-helpers/testing';

import { UsersSelectors } from '../+state/users.selectors';
import { MockUsersSelectors } from '../testing';
import { UserDetailComponent } from './user-detail.component';

describe('UserDetailComponent', () => {
  let fixture: ComponentFixture<UserDetailComponent>;

  beforeEach(
    async(() => {
      TestBed.configureTestingModule({
        declarations: [UserDetailComponent],
        schemas: [NO_ERRORS_SCHEMA],
        providers: [
          { provide: UsersSelectors, useClass: MockUsersSelectors },
          { provide: Store, useValue: MockStore }
        ]
      }).compileComponents();
    })
  );

  beforeEach(() => {
    fixture = TestBed.createComponent(UserDetailComponent);
    fixture.detectChanges();
  });

  it('should render user form', () => {
    expect(fixture).toMatchSnapshot();
  });
});
