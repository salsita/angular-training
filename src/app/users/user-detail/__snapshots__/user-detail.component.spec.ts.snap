// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetailComponent should render user form 1`] = `
<app-user-detail
  skills$={[Function Observable]}
  store={[Function Function]}
  user$={[Function Observable]}
>
  <app-user-form />
  
  
</app-user-detail>
`;
